node/src/components/deploy_buffer.rs

```diff
@@ -224,47 +258,72 @@ impl DeployBuffer {
     {
-        // TODO: This will ignore deploys that weren't in `collected_deploys`. They might be added
+        // TODO: This will ignore deploys that weren't in `pending`. They might be added
         // later, and then would be proposed as duplicates.
         let deploy_map: HashMap<_, _> = deploys
             .into_iter()
             .filter_map(|deploy_hash| {
-                self.collected_deploys
+                self.pending
                     .get(&deploy_hash)
                     .map(|deploy| (deploy_hash, deploy.clone()))
             })
             .collect();
-        self.collected_deploys
+        self.pending
             .retain(|deploy_hash, _| !deploy_map.contains_key(deploy_hash));
-        self.processed.insert(block, deploy_map);
+        self.proposed.insert(block, deploy_map);
     }
 
     /// Notifies the deploy buffer that a block has been finalized.
     fn finalized_block(&mut self, block: ProtoBlockHash) {
-        if let Some(deploys) = self.processed.remove(&block) {
-            self.collected_deploys
+        if let Some(deploys) = self.proposed.remove(&block) {
+            self.pending
                 .retain(|deploy_hash, _| !deploys.contains_key(deploy_hash));
             self.finalized.insert(block, deploys);
         } else if !block.is_empty() {
             // TODO: Events are not guaranteed to be handled in order, so this could happen!
-            error!("finalized block that hasn't been processed!");
+            error!("finalized block that hasn't been proposed!");
         }
     }
 
     /// Notifies the deploy buffer that a block has been orphaned.
     fn orphaned_block(&mut self, block: ProtoBlockHash) {
-        if let Some(deploys) = self.processed.remove(&block) {
-            self.collected_deploys.extend(deploys);
+        if let Some(deploys) = self.proposed.remove(&block) {
+            self.pending.extend(deploys);
         } else {
             // TODO: Events are not guaranteed to be handled in order, so this could happen!
-            error!("orphaned block that hasn't been processed!");
+            error!("orphaned block that hasn't been proposed!");
+        }
+    }
+
+    /// Prunes expired deploy information from the DeployBuffer, returns the total deploys pruned
+    fn prune(&mut self, current_instant: Timestamp) -> usize {
+        fn prune_deploys(deploys: &mut DeployCollection, current_instant: Timestamp) -> usize {
+            let initial_len = deploys.len();
+            deploys.retain(|_hash, header| !header.expired(current_instant));
+            initial_len - deploys.len()
+        }
+        fn prune_blocks(blocks: &mut ProtoBlockCollection, current_instant: Timestamp) -> usize {
+            let mut pruned = 0;
+            let mut remove = Vec::new();
+            for (block_hash, proposed) in blocks.iter_mut() {
+                pruned += prune_deploys(proposed, current_instant);
+                if proposed.is_empty() {
```

Comment:
I hate Rust for this… `proposed` is mutated inside the `prune_deploys`. Referential Transparency and immutability FTW.

Reply:
Ah that's just a stale binding name :( I'll fix that